diff --git a/jellyfish/dump_main.cc b/jellyfish/dump_main.cc
index 187d29d..0620b0c 100644
--- a/jellyfish/dump_main.cc
+++ b/jellyfish/dump_main.cc
@@ -38,13 +38,13 @@ void dump(iterator& it, std::ostream &out,
     while(it.next()) {
       if(it.get_val() < lower_count || it.get_val() > upper_count)
         continue;
-      out << it.get_dna_str() << spacer << it.get_val() << "\n";
+      out << it.get_key() << spacer << it.get_val() << "\n";
     }
   } else {
     while(it.next()) {
       if(it.get_val() < lower_count || it.get_val() > upper_count)
         continue;
-      out << ">" << it.get_val() << "\n" << it.get_dna_str() << "\n";
+      out << ">" << it.get_val() << "\n" << it.get_key() << "\n";
     }
   }
 }
